### PR TITLE
[flutter_tools] well known device ids

### LINKED ISSUE #83115: Special case macos, linux, windows, winuwp, chrome device identifiers in device discovery
This would allow us to avoid running other device discoverers, which would both be faster and avoid errors from attached but not available device types.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Currently the flutter tool will always run all device discovery code, even if the id provided corresponds to a "well-known" device such as "chrome" or "windows". While the device discovery is stopped as soon as a device is identified, this still leaves room for extra logs and the potential for errors in device discovery to leak out.

Instead, allow each device discoverer to declare a list of identifiers that are treated as "well known". Then, if the device discovery was provided a specific id and that id is contained in the set, only run discovery once.

This does not apply in the reverse direction; since there is almost no cost to device discovery for the "well known" devices, skipping it when a not well known identifier is provided does not buy much.

Fixes https://github.com/flutter/flutter/issues/83115

### CODE REVIEW

**Review by 3zrv — APPROVED** [review]

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart [hunk 1]
```diff
@@ -268,9 +267,6 @@ class FlutterTesterDevices extends PollingDeviceDiscovery {
     @required ProcessManager processManager,
     @required Logger logger,
     @required FlutterVersion flutterVersion,
-    // TODO(jonahwilliams): remove after flutter rolls
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/tester/flutter_tester.dart:271]: While I'm here

File: packages/flutter_tools/lib/src/device.dart [hunk 2]
```diff
@@ -336,6 +343,15 @@ abstract class DeviceDiscovery {
   /// Gets a list of diagnostic messages pertaining to issues with any connected
   /// devices (will be an empty list if there are no issues).
   Future<List<String>> getDiagnostics() => Future<List<String>>.value(<String>[]);
+
+  /// The well-known device IDs that the discoverer can produce.
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/device.dart:347]: Maybe "The well-known, hard-coded device IDs that the discoverer can produce." it's defining well-known as "well-known".  See: my confusion last week with `ephemeral`
https://github.com/flutter/flutter/blob/d4be5726389869e7843c02337189aa7e5386ef8c/packages/flutter_tools/lib/src/device.dart#L436-L437

File: packages/flutter_tools/lib/src/macos/macos_ipad_device.dart [hunk 3]
```diff
@@ -144,4 +144,7 @@ class MacOSDesignedForIPadDevices extends PollingDeviceDiscovery {
 
   @override
   Future<List<String>> getDiagnostics() async => const <String>[];
+
+  @override
+  List<String> get wellKnownIds => const <String>[];
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/macos/macos_ipad_device.dart:149]: Maybe this ID should be put in a constant?
```suggestion
  List<String> get wellKnownIds => const <String>['designed-for-ipad'];
```

File: packages/flutter_tools/lib/src/device.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/device.dart]: // This is a comment.

I updated this so that well known is specified  as hard-coded ids

**Review by jmagman — APPROVED** [review]
LGTM!

**Pull request merged**